### DISCUSSION TITLE
Random Forest Completed - 3 ML Stage Milestone

### DISCUSSION THREAD

**Discussion by saunakdas16** [saunakdas16/alzheimers-eeg-ml#3] [Announcements]
# 🌲 Machine Learning Milestone -  Random Forest Completed

Another major milestone has been completed in my **EEG-Based Alzheimer's Disease Classification** project.

After completing the EEG preprocessing, artifact correction, ICA, spectral analysis, and automated subject-level feature extraction across the complete dataset of **88 participants**, I have now completed the third machine-learning classification experiment: **Random Forest**.

The current machine-learning stage now includes:

**Logistic Regression → SVM → Random Forest**

## 🌲 Random Forest

The Random Forest experiment used the same **65-participant AD vs Cognitively Normal (CN)** dataset and the same stratified **80/20 train-test split** used for the previous models:

- **36 AD participants**
- **29 CN participants**
- **52 training participants**
- **13 held-out test participants**

The model used five subject-level EEG relative spectral-power features:

- Delta
- Theta
- Alpha
- Beta
- Gamma

Unlike Logistic Regression and SVM, Random Forest does not require feature standardization because it is a tree-based model.

## 📊 Initial Results

The initial Random Forest achieved:

**Test Accuracy:** 84.6%  
**ROC-AUC:** 0.857

On the held-out test set, **11 of 13 participants were correctly classified**.

The confusion matrix showed:

```text
                 Predicted
                 AD    CN

Actual AD         5     2
Actual CN         0     6
```

## 🔧 Cross-Validation & Hyperparameter Tuning

A **5-fold stratified cross-validation** procedure was performed only within the **52 training participants**.

The initial cross-validation results were:

**Mean CV Accuracy:** 77.5%  
**Standard Deviation:** 13.2%

Hyperparameter tuning was then performed using `GridSearchCV`.

The search evaluated:

```text
n_estimators = 100, 200, 300
max_depth = None, 5, 10
min_samples_split = 2, 5
min_samples_leaf = 1, 2
```

This resulted in **36 parameter combinations**, each evaluated using 5-fold stratified cross-validation on the training set.

The selected configuration was:

```text
n_estimators = 100
max_depth = 5
min_samples_split = 2
min_samples_leaf = 1
```

The best mean cross-validation accuracy was:

**79.3%**

with a cross-validation standard deviation of:

**10.2%**

## ✅ Tuned Random Forest

The tuned Random Forest was evaluated on the same untouched **13-participant test set**.

**Test Accuracy:** 84.6%  
**ROC-AUC:** 0.857

The tuned model produced the same class predictions as the initial Random Forest, so the final test-set accuracy remained **84.6%**.

## 📈 Model Visualization

The Random Forest analysis now includes:

**Prediction Confidence → Feature Interpretation → Feature Contributions**

A new **Random Forest prediction-confidence visualization** shows the predicted CN probability for each test participant, the 0.5 decision threshold, actual and predicted classes, and correct versus incorrect predictions.

A **permutation-based feature-importance visualization** was also created to examine how much model accuracy changes when individual EEG features are shuffled.

A **feature-contribution heatmap** was created by comparing the model's predicted CN probability before and after replacing individual test-set features with their corresponding training-set means.

## 🔒 Methodological Considerations

As with the previous models, the final **13-participant test set was kept separate from model selection and hyperparameter tuning**.

Cross-validation and parameter selection for Random Forest were performed only within the **52 training participants**.

This maintains the separation between training/model-selection data and the final held-out test set.

Because the test set contains only **13 participants**, the reported test-set metrics should be interpreted as results for this particular experimental split rather than as evidence of clinical diagnostic performance or generalization to an independent population.

## 🧠 Three-Model Classification Stage Completed

With the completion of Random Forest, the initial three-model classification stage is now complete:

**Logistic Regression**  
**Support Vector Machine (SVM)**  
**Random Forest**

The project now contains a consistent subject-level machine-learning workflow together with model-specific interpretation and visualization.

## 🔮 What's Next?

The future steps will focus on consolidating the results from all three models, examining their behavior and EEG feature patterns more systematically, and documenting the findings within the project.

This continues to be both a **research-oriented EEG analysis project and a learning journey in computational neuroscience, signal processing, and machine learning**.

🔗 **Repository:**  
https://github.com/saunakdas16/alzheimers-eeg-ml
